river/ensemble/boosting.py

```diff
@@ -119,3 +124,189 @@ def predict_proba_one(self, x):
+            if correct:
+                pos = self.orderPosition[maxAcc]
+                maxAcc += 1
+            else:
+                pos = self.orderPosition[minAcc]
+                minAcc -= 1
+
+            for _ in range(poisson(lambda_poisson, self._rng)):
+                models[pos][1].learn_one(x, y)
+
+            if models[pos][1].predict_one(x) == y:
+                self.correct_weight[pos] += lambda_poisson
+                lambda_poisson *= (self.instances_seen) / (2 * self.correct_weight[pos])
+                correct = True
+            else:
+                self.wrong_weight[pos] += lambda_poisson
+                lambda_poisson *= (self.instances_seen) / (2 * self.wrong_weight[pos])
+                correct = False
+        return self
+
+    def predict_proba_one(self, x):
+        y_proba = collections.Counter()
+        for i, model in enumerate(self):
+            model_weight = 0.0
+            if self.correct_weight[i] > 0.0 and self.wrong_weight[i] > 0.0:
+                epsilon = self.wrong_weight[i] / (self.correct_weight[i] + self.wrong_weight[i])
+                if epsilon <= self.error_bound:
+                    beta_inv = (1 - epsilon) / epsilon
+                    model_weight = np.log(beta_inv)
+
+            if model_weight != 0.0:
+                predictions = model.predict_proba_one(x)
+                normalize_values_in_dict(predictions, inplace=True)
+                scale_values_in_dict(predictions, model_weight, inplace=True)
+                y_proba.update(predictions)
+
+        normalize_values_in_dict(y_proba, inplace=True)
+        return y_proba
+
+
+class BOLEBaseModel(base.Classifier):
+    """BOLEBaseModel
+
+    In case a warning is detected, a background model starts to train. If a drift is detected, the model will be replaced by the background model.
+
+    Parameters
+    ----------
+    model
+        The classifier and background classifier class.
+    drift_detection_method
+        Algorithm to track warnings and concept drifts
+    """
+
+    def __init__(self, model: base.Classifier, drift_detection_method: DriftDetector):
+        self.base_class = model
+        self.model = copy.deepcopy(self.base_class)
+        self.bkg_model = None
+        self.drift_detection_method = (
+            drift_detection_method if drift_detection_method is not None else DDM()
+        )
+
+    def predict_proba_one(self, x):
+        return self.model.predict_proba_one(x)
+
+    def learn_one(self, x, y):
+        self.update_ddm(x, y)
+        return self.model.learn_one(x, y)
+
+    def update_ddm(self, x, y):
+        y_pred = self.model.predict_one(x)
+        if y_pred is not None:
+            incorrectly_classifies = int(y_pred != y)
+            self.drift_detection_method.update(incorrectly_classifies)
+            if self.drift_detection_method.warning_detected:
+                if self.bkg_model is None:
+                    self.bkg_model = copy.deepcopy(self.base_class)
```

Comment:
I think you can do `self.bkg_model = copy.deepcopy(self.base_class)` (or `self.base_class.clone()`) in the `__init__` (on line 286) and then delete this `if`. Then you can call `self.bkg_model = copy.deepcopy(self.base_class)` (or `self.base_class.clone()`) at line 310

Reply:
But maybe I'm wrong! I'll let you check :)

Reply:
I use None to avoid training without having a model (and maybe avoid some bugs).